app/javascript/src/Policies/components/CustomPolicy.jsx

```diff
@@ -0,0 +1,143 @@
+      ? schema.configuration
+      : (function () { throw new Error('Policy configuration not found') })()
+  }
+
+  const onCodeChange = (editor, metadata, code) => {
+    try {
+      onChange(getConfiguration(fromJson(code)))
+      setState({ valid: true, code })
+    } catch (err) {
+      console.warn(err)
+      setState({ valid: false, code })
+    }
+  }
+
+  const [ icon, cls, hiddenClass, errorClass ] = state.valid
+    ? [ 'check', 'valid', 'is-hidden', '' ]
+    : [ 'times', 'invalid', '', 'CustomPolicy-error' ]
+
+  return (
+    <div className={`${errorClass} panel panel-default`}>
+      <div className="panel-heading">
+        <i className={`${cls} fa fa-${icon}`} />
+        {' JSON Schema'}
+      </div>
+      <CodeMirror
+        value={state.code}
+        onChange={onCodeChange}
+        autoCursor={false}
+        options={CM_OPTIONS}
+      />
+      <div className={hiddenClass}> There's an error in the JSON Schema</div>
+    </div>
+  )
+}
+
+function CustomPolicyForm ({policy, onChange}: {policy: Policy, onChange: OnChange}): React.Node {
+  const isNewPolicy = !(policy.id && policy.id !== 0)
+  const action = (isNewPolicy) ? '/p/admin/registry/policies/' : `/p/admin/registry/policies/${policy.id}`
+  return (
+    <form action={action} method="post">
+      <label>
+        Directory:
+        <input type="text" name="directory" value={policy.directory} onChange={onChange} disabled={!isNewPolicy} />
+      </label>
+      <input type="hidden" name="id" value={policy.id} disabled={true} />
+      <input name="schema" type="hidden" value={JSON.stringify(policy.schema)} />
+      {(!isNewPolicy) ? <input name="_method" type="hidden" value='put' /> : ''}
+      <input type="submit" />
+      <CSRFToken />
+    </form>
+  )
+}
+
+function Form ({policy}: {policy: Policy}): React.Node {
+  const [pol, setPolicy] = useState(policy)
+  const onSchemaEdited = (pol: Policy) => (configuration: Object) => setPolicy({...pol, ...{schema: {...pol.schema, ...{configuration}}}})
+  const handleChange = (pol: Policy) => (ev: InputEvent) => setPolicy({...pol, ...{[ev.target.name]: ev.target.value}})
+
+  return (
+    <div>
+      <div className="CustomPolicy-editor">
+        <Editor className="CustomPolicy-code" code={pol.schema} onChange={onSchemaEdited(pol)} />
+        <SchemaForm className="CustomPolicy-form" schema={pol.schema.configuration} />
+      </div>
+      <CustomPolicyForm policy={pol} onChange={handleChange(pol)} />
+    </div>
+  )
+}
+
+function CustomPolicy ({policy = POLICY_TEMPLATE}: {policy: Policy}): React.Node {
+  return (
+    <section className="CustomPolicy">
+      <header className='CustomPolicy-header'>
+        <h2 className="CustomPolicy-title">Custom Policy</h2>
+        <a className="CustomPolicy-cancel" href="/p/admin/registry/policies" >
+          <i className="fa fa-times-circle" /> Cancel
+        </a>
+      </header>
+      <Form policy={policy} />
+    </section>
+  )
+}
+
+export {
```

Comment:
It's good when you are re-using those components. In this case, we only are using them inside CustomPolicy. It happens with PolicyTile and List, they needed to be refactored to different files because they were reused by others.

Reply:
I disagree, for the following reasons:
- Even if it's not reused now that doesn't mean it won't eventually. It won't be _needed to be refactored to diff files_.
- It's less maintainable and harder to read and understand. With smaller pieces it's easier to find errors and gaps in implementation, also to end up with a well defined logic and components.
- It breaks the SRP
- Specs will also be a huge file of mixed test cases. It's far easier to define specs when there is less functionality to test.
- It's simply tidier

Reply:
1. That logic is totally wrong. You just don't go creating modules because "You might use them in the future." You are definitely not creating a Framework before you need it.
2. It will evolve when it's needed, and about maintainability, it's in the same file, you don't need to go to other files for something which is used solely there.
3. If you say it breaks it because I'm exposing the rest, it's just for testing purposes only.
4. When ever it needs to be reused, it will be separated.
5. that's only your opinion.
3.

Reply:
It breaks SRP in the sense that the same file contains many objects with different purpose:
There's the parent `Form` that controls some state and overall structure.
Then the `Editor` that is textbox with format and validation and, also using an external dependency.
`CustomPolicyForm` I fail to see what it does.
`CustomPolicy` apparently is a different thing, not used in or using Form.

I am not trying to convince you but I want to understand why you're reluctant to have smaller pieces of logic that are _easier_ to read and maintain. Or at a justification why my arguments are not valid, regardless of personal preferences.

Reply:
Think about CustomPolicy as the sole API method you are using and the others as private methods. You might want to test them, but not use them anywhere else

Reply:
I suggest you to watch this https://www.youtube.com/watch?v=XpDsk374LDE